scipy/stats/stats.py

```diff
@@ -2465,15 +2465,20 @@ def gstd(a, axis=0, ddof=1):
             raise ValueError(
                 'Infinite value encountered. The geometric standard deviation '
                 'is defined for strictly positive values only.')
-        elif np.less_equal(a, 0).any():
+        a_nan = np.isnan(a)
+        a_nan_any = a_nan.any()
+        if a_nan.any():
+            return np.exp(np.std(log(a, where=~a_nan), axis=axis, ddof=ddof))
```

Comment:
The check for nan should go after the less equal 0 check. Otherwise the type of exception for negative data will depend on whether a nan is also present.